File: backend/src/main/java/org/luxons/sevenwonders/game/moves/PlayFreeCardMove.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.cards.Card;

public class PlayFreeCardMove extends CardFromHandMove {

    PlayFreeCardMove(int playerIndex, Card card, PlayerMove move) {
        super(playerIndex, card, move);
    }

    @Override
    public void validate(Table table, List<Card> playerHand) throws InvalidMoveException {
        super.validate(table, playerHand);
        Board board = table.getBoard(getPlayerIndex());
        if (!board.canPlayFreeCard(table.getCurrentAge())) {
            throw new InvalidMoveException(
                    String.format("Player %d cannot play the card %s for free", getPlayerIndex(), getCard().getName()));
        }
    }

    @Override
    public void place(Table table, List<Card> discardedCards, Settings settings) {
        Board board = table.getBoard(getPlayerIndex());
        board.addCard(getCard());
    }

    @Override
    public void activate(Table table, List<Card> discardedCards, Settings settings) {
        // only apply effects, without paying the cost
        getCard().getEffects().forEach(e -> e.apply(table, getPlayerIndex()));
        Board board = table.getBoard(getPlayerIndex());
        board.consumeFreeCard(table.getCurrentAge());
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/PlayFreeCardMove.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.Settings
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.api.Table
import org.luxons.sevenwonders.game.boards.Board
import org.luxons.sevenwonders.game.cards.Card

class PlayFreeCardMove internal constructor(playerIndex: Int, card: Card, move: PlayerMove) :
    CardFromHandMove(playerIndex, card, move) {

    @Throws(InvalidMoveException::class)
    override fun validate(table: Table, playerHand: List<Card>) {
        super.validate(table, playerHand)
        val board = table.getBoard(playerIndex)
        if (!board.canPlayFreeCard(table.currentAge)) {
            throw InvalidMoveException(
                String.format("Player %d cannot play the card %s for free", playerIndex, card.name)
            )
        }
    }

    override fun place(table: Table, discardedCards: MutableList<Card>, settings: Settings) {
        val board = table.getBoard(playerIndex)
        board.addCard(card)
    }

    override fun activate(table: Table, discardedCards: List<Card>, settings: Settings) {
        // only apply effects, without paying the cost
        card.effects.forEach { e -> e.apply(table, playerIndex) }
        val board = table.getBoard(playerIndex)
        board.consumeFreeCard(table.currentAge)
    }
}
